src/ServiceControl.Monitoring/QueueLength/LegacyQueueLengthReportHandler.cs

```diff
@@ -9,20 +10,47 @@
 
     class LegacyQueueLengthReportHandler : IHandleMessages<MetricReport>
     {
-        ConcurrentDictionary<string, string> loggedInstances = new ConcurrentDictionary<string, string>();
+        LegacyQueueLengthEndpoints legacyEndpoints;
+
+        public LegacyQueueLengthReportHandler(LegacyQueueLengthEndpoints legacyEndpoints)
+        {
+            this.legacyEndpoints = legacyEndpoints;
+        }
 
         public Task Handle(MetricReport message, IMessageHandlerContext context)
         {
             var endpointInstanceId = EndpointInstanceId.From(context.MessageHeaders);
 
-            if (loggedInstances.TryAdd(endpointInstanceId.InstanceId, endpointInstanceId.InstanceId))
+            if (legacyEndpoints.TryAdd(endpointInstanceId.InstanceId))
             {
                 Logger.Warn($"Legacy queue length report received from {endpointInstanceId.InstanceName} instance of {endpointInstanceId.EndpointName}");
             }
 
             return TaskEx.Completed;
         }
 
+        public class LegacyQueueLengthEndpoints
+        {
+            static TimeSpan cleanInterval = TimeSpan.FromHours(1);
+
+            ConcurrentDictionary<string, string> regsteredInstances = new ConcurrentDictionary<string, string>();
+            DateTime lastCleanTime = DateTime.UtcNow;
+
+            public bool TryAdd(string id)
+            {
+                var now = DateTime.UtcNow;
+
+                if (lastCleanTime.Add(cleanInterval) < now)
+                {
+                    lastCleanTime = now;
```

Comment:
I'm sure that because of the operations on the `registeredInstances` this won't happen, but still, I want to notice that a field is written without any lock. This means that depending on the impl of copying this struct, the assignment might not be atomic and one could see a partially written value.

Reply:
I've switched to `long` as the field type and added `Interlocked` for setting the value.